src/functions.php

```diff
@@ -110,6 +110,40 @@ function is_first_key_operator($document)
     return (isset($firstKey[0]) && $firstKey[0] === '$');
 }
 
+/**
+ * Returns whether an update specification is a valid aggregation pipeline.
+ *
+ * @internal
+ * @param mixed $pipeline
+ * @return boolean
+ */
+function is_pipeline($pipeline)
+{
+    if (! is_array($pipeline)) {
+        return false;
+    }
+
+    if ($pipeline === []) {
```

Comment:
Continuing from my previous shell testing, it looks like MongoDB 4.2 accepts an empty pipeline and simply NOPs:

```
rs0:PRIMARY> db.runCommand({update:"foo",updates:[{q:{},u:[]}]});
{
	"n" : 1,
	"nModified" : 0,
        ...
}
```

Contrast this with an empty replacement document:

```
rs0:PRIMARY> db.runCommand({update:"foo",updates:[{q:{},u:{}}]});
{
	"n" : 1,
	"nModified" : 1,
	...
}
```

Is this function based on libmongoc's [`_mongoc_document_is_pipeline()`](https://github.com/mongodb/mongo-c-driver/blob/1.15.0/src/libmongoc/src/mongoc/mongoc-util.c#L530)? I see that also returns `false` if the argument has no elements.

I'm not arguing that this is incorrect. I actually think it's the most intuitive behavior, especially for PHP where users would most commonly pass `[]` as an empty replacement document; however, I'm curious if this was dictated by the spec. If not, perhaps we should mention it in the SPEC ticket I proposed opening in my earlier comment.

Reply:
> Is this function based on libmongoc's [`_mongoc_document_is_pipeline()`](https://github.com/mongodb/mongo-c-driver/blob/1.15.0/src/libmongoc/src/mongoc/mongoc-util.c#L530)? I see that also returns `false` if the argument has no elements.

That is correct. I assumed that it'd be wrong to treat an empty array as an aggregation pipeline, but I also specifically remember one test failing because of that, which is why I added the check after all.